src/misc/ethernet_slirp.cpp

```diff
@@ -0,0 +1,267 @@
+}
+
+void slirp_notify(void *opaque)
+{
+	(void)opaque;
+	return;
+}
+
+SlirpEthernetConnection::SlirpEthernetConnection(void)
+      : EthernetConnection()
+{
+	slirp_callbacks.send_packet = slirp_send_packet;
+	slirp_callbacks.guest_error = slirp_guest_error;
+	slirp_callbacks.clock_get_ns = slirp_clock_get_ns;
+	slirp_callbacks.timer_new = slirp_timer_new;
+	slirp_callbacks.timer_free = slirp_timer_free;
+	slirp_callbacks.timer_mod = slirp_timer_mod;
+	slirp_callbacks.register_poll_fd = slirp_register_poll_fd;
+	slirp_callbacks.unregister_poll_fd = slirp_unregister_poll_fd;
+	slirp_callbacks.notify = slirp_notify;
+}
+
+SlirpEthernetConnection::~SlirpEthernetConnection(void)
+{
+	if(slirp) slirp_cleanup(slirp);
+}
+
+bool SlirpEthernetConnection::Initialize(Section* dosbox_config)
+{
+	(void)dosbox_config; /* Ignore for now */
+
+	/* Config */
+	config.version = 1;
+	config.restricted = 0; /* Allow access to host */
+	config.disable_host_loopback = 0; /* Allow access to 127.0.0.1 */
+	config.if_mtu = 0; /* IF_MTU_DEFAULT */
+	config.if_mru = 0; /* IF_MRU_DEFAULT */
+	config.enable_emu = 0; /* Buggy, don't use */
+	/* IPv4 */
+	config.in_enabled = 1;
+	inet_pton(AF_INET, "10.0.2.0", &config.vnetwork);
+	inet_pton(AF_INET, "255.255.255.0", &config.vnetmask);
+	inet_pton(AF_INET, "10.0.2.2", &config.vhost);
+	inet_pton(AF_INET, "10.0.2.3", &config.vnameserver);
+	inet_pton(AF_INET, "10.0.2.15", &config.vdhcp_start);
+	/* IPv6 code is left here as reference but disabled as no DOS-era
+	 * software supports it and might get confused by it */
+	config.in6_enabled = 0;
+	inet_pton(AF_INET6, "fec0::", &config.vprefix_addr6);
+	config.vprefix_len = 64;
+	inet_pton(AF_INET6, "fec0::2", &config.vhost6);
+	inet_pton(AF_INET6, "fec0::3", &config.vnameserver6);
+	/* DHCPv4, BOOTP, TFTP */
+	config.vhostname = "DOSBox-X";
+	config.vdnssearch = NULL;
+	config.vdomainname = NULL;
+	config.tftp_server_name = NULL;
+	config.tftp_path = NULL;
+	config.bootfile = NULL;
+
+	slirp = slirp_new(&config, &slirp_callbacks, this);
+	if(slirp)
+	{
+		LOG_MSG("SLIRP successfully initialized");
+		return true;
+	}
+	else
+	{
+		/* TODO: better error message? */
+		LOG_MSG("SLIRP failed to initialize");
+		return false;
+	}
+}
+
+void SlirpEthernetConnection::SendPacket(uint8_t* packet, int len)
+{
+	slirp_input(slirp, packet, len);
+}
+
+void SlirpEthernetConnection::GetPackets(std::function<void(uint8_t*, int)> callback)
+{
+	get_packet_callback = callback;
+	uint32_t timeout = 0;
+	Polls_Clear();
+	slirp_pollfds_fill(slirp, &timeout, slirp_add_poll, this);
+	/* TODO: replace poll with select */
+	int ret = poll(polls, 256, timeout);
+	slirp_pollfds_poll(slirp, (ret <= -1), slirp_get_revents, this);
+	Timers_Run();
+}
+
+void SlirpEthernetConnection::Receive_Packet(uint8_t* packet, int len)
+{
+	get_packet_callback(packet, len);
+}
+
+struct slirp_timer* SlirpEthernetConnection::Timer_New(SlirpTimerCb cb, void *cb_opaque)
+{
+	for(int i = 0; i < 256; ++i)
+	{
```

Comment:
Regarding the 256 magic number quality of timers and loops, suggest it be defined as a private `costexpr int NUM_TIMERS = 256;`.

Also, where does the 256 come from? If it's tied to number of open file-descriptions (or connections), maybe the `timer[]` array could be a `std::vector` or `std:deque`, and the timers dynamically `push`ed and `pop`ed, to avoid always looping over 256 of them. (std:deque is also very fast).

Reply:
> I just made it up. I'm going to be replacing this with a proper system for dynamically allocating them. I can't use fancy C++ containers for them because they don't let you access the underlying array (except in newer C++ versions)

The underlying `std::vector` array `.data()` is guaranteed to be the contiguous series elements (just as if you used a new[] array of them).